### PR TITLE
Add object orientated interface

### PR DESCRIPTION
**Pull request by AnthonyHorton**
Putting this in as a draft PR to request feedback. While reviewing a PR on a project that uses the `sn3218` library I got inspired to create an alternative, object orientated interface. This might appeal to users who are accustomed to idiomatic Python but unfamiliar with things like bitmasks.

I've left the original interface as is (moved, but still accessible via `import sn3218`) and added an approximately functionally equivalent but more Pythonic interface based on a new class `SN3218`. Example:

```
from sn3218.sn3218class import SN3218

# Automation HAT LED names & numbers
led_names = {'ANALOG_IN_1': 1,
             'ANALOG_IN_2': 2,
             ...
             'WARN': 16,
             'COMMS': 17,
             'POWER': 18}

leds = SN3218(led_names)  # led_names is optional, default names ('ONE', 'TWO',...) also available
leds.leds_brightness = 32
leds.set_leds_brightness({'POWER': 64,})
leds.turn_on_leds(['POWER', 'WARN'])
leds.set_leds_enabled({'ANALOG_IN_2': True, 'NINE': False, 12: True})
leds.set_led_gamma('ANALOG_IN_2', new_gamma_table)
leds.reset_led_gamma('ANALOG_IN_2')
leds.turn_off_leds()

leds.demo()  # Reproduces behaviour of original self test/demo script
```

The input/outputs are mostly dictionaries or lists of strings and ints. The bit twiddling is all handled internally, using `enum.Flag`. All the public methods and properties are documented with full docstrings.

I'm aware that my changes have probably broken all the packaging stuff so that would need fixing before this could be merged, but I wanted to find out whether there is in fact any interest in merging this before I put the time in to fix the packaging. In the meantime `python setup.py develop` works fine for testing purposes. Let me know what you think.

### CODE REVIEW

**Comment by Gadgetoid** [comment]
Thanks for taking the time to make this and raise it as a PR.

I like the concept in principle, but I'm hesitant to make any far-reaching changes to this library since it's pretty stable (at least it hasn't caused me any recent headaches) and it's used in a handful of our product libraries as a dependency. (and I haven't - yet - pinned them against a specific version)

Possibly the biggest issue with merging this code is that it's biased toward Python 3.4+ whereas we have to continue support for 2.7 due to a combination of it being the default `python` on the Raspberry Pi, and the average user not understanding or caring that it's been superceded by 3.x.

I believe it's possible to package in such a way that this class-based interface is available only in the Python 3.x build of the library. That's a possible consideration. And I don't think anything I've said necessarily precludes us from merging - it just requires time, which I'm invariably short on.

Your LED naming approach solves a very similar problem to the one I'm tackling in `i2cdevice`- you might want to investigate the use of `*args` and `**kwargs` to remove some of your syntactic complexity.

For instance, I believe:

* `leds.set_leds_brightness({'POWER': 64,})` could become `leds.set_brightness(power=64)`
* `leds.turn_on_leds(['POWER', 'WARN'])` could become `leds.turn_on('power', 'warn')`

Also

```python
leds.set_leds_enabled({'ANALOG_IN_2': True, 'NINE': False, 12: True})
```

could become:

```python
leds.set_enabled(analog_in_2=True, nine=False)
```

While this pattern isn't compatible with using LED indexes, there's no reason not to support both, or make a clearer separation between the two:

```python
leds.enable('analog_in', 12)
leds.disable('nine')
```

To throw another spanner into the works- I've seen (and I'm not sure I'm totally qualified to judge) "Pythonic" class-based libraries for dealing with individual things use a class-per-item approach, which would make the `sn3218` driver work a bit like this:

```
import sn3218
warn = sn3218.LED(12, brightness=64, gamma=[...])  # Enables LED
warn.on()
warn.off()
```

This would depend upon the sn3218 hardware to maintain the global state for all the LEDs, which is a bit messy. Not to mention you can end up with 18 class instances to handle all your LEDs.

GPIOZero is probably the canonical example of this on the Pi: https://gpiozero.readthedocs.io/en/stable/

## So TLDR:

I don't want our libraries to be *too* bounded by our own business/hardware/library requirements so I'm definitely up for entertaining the idea of merging this so that others can take advantage of the more descriptive API.

* Look into `*args` and `**kwargs` to refine the API- maybe this is a bad idea, I'm not entirely sure
* Look into packaging the class-based approach for Python 3.x only

**Comment by AnthonyHorton** [comment]
Thanks for the detailed (and rapid) feedback!

Yep, I appreciate the importance of maintaining the existing interface exactly as is in order to not break dependent packages. I was attempting to do that, and I believe that these changes do leave it working exactly as before. It's just the packaging that I'll have (temporarily) broken due to moving some files around.

Your `*args, **kwargs` suggestion looks good, at least at first glance. The examples you show are certainly less clunky than passing in lists and dictionaries. I'll have a go at doing something like that.

As for the Python 2/3 issue, I'm happy to wall this off as a Python 3 only thing. I moved on from Python 2 years ago and I don't think I could bring myself to sink much effort into making this Python 2 compatible...

Right, I'll do some polishing when I get time and mark is ready for review when I'm done.

**Comment by Gadgetoid** [comment]
Admittedly I'm probably being overly cautious about dependent packages- but in my experience it's better to be safe than sorry.

I'd love to leave Python 2.x behind! Hopefully we can slowly guide people toward 3.x with tweaks to documentation and examples. That said, I can't break the habit of running in 2.x myself!

Looking forward to looking over your ready-for-review PR!

**Comment by Gadgetoid** [comment]
This library has been deprecated in favour of https://github.com/pimoroni/sn3218-python which implements an OO class interface, granted not one with a featureset this comprehensive. Would be happy to consider a new PR if you're interested, but no worries either way, I think we've all had a bit of a moment over the intervening years since 2019 :scream:
